File: csw-event-impl/src/test/scala/csw/services/event/internal/redis/PubSubTest.scala
package csw.services.event.internal.redis

import akka.actor.ActorSystem
import com.github.sebruck.EmbeddedRedis
import csw.messages.commons.CoordinatedShutdownReasons.TestFinishedReason
import csw.messages.events.EventKey
import csw.services.event.helpers.RegistrationFactory
import csw.services.event.helpers.TestFutureExt.RichFuture
import csw.services.event.helpers.Utils.makeEvent
import csw.services.event.internal.EventServicePubSubTestFramework
import csw.services.event.internal.commons.{EventServiceConnection, Wiring}
import csw.services.event.scaladsl.RedisFactory
import csw.services.location.commons.ClusterAwareSettings
import csw.services.location.scaladsl.LocationServiceFactory
import io.lettuce.core.{RedisClient, RedisURI}
import org.scalatest.{BeforeAndAfterAll, FunSuite, Matchers}
import redis.embedded.RedisServer

//DEOPSCSW-334: Publish an event
//DEOPSCSW-335: Model for EventName that encapsulates the topic(or channel ) name
//DEOPSCSW-337: Subscribe to an event based on prefix
class PubSubTest extends FunSuite with Matchers with BeforeAndAfterAll with EmbeddedRedis {
  private val seedPort        = 3558
  private val redisPort       = 6379
  private val clusterSettings = ClusterAwareSettings.joinLocal(seedPort)
  private val locationService = LocationServiceFactory.withSettings(ClusterAwareSettings.onPort(seedPort))
  private val tcpRegistration = RegistrationFactory.tcp(EventServiceConnection.value, redisPort)
  locationService.register(tcpRegistration).await

  private val redis = RedisServer.builder().setting(s"bind ${clusterSettings.hostname}").port(redisPort).build()

  private implicit val actorSystem: ActorSystem = clusterSettings.system
  private val redisClient                       = RedisClient.create()
  private val wiring                            = new Wiring(actorSystem)
  private val redisFactory                      = new RedisFactory(redisClient, locationService, wiring)
  private val publisher                         = redisFactory.publisher().await
  private val subscriber                        = redisFactory.subscriber().await
  private val framework                         = new EventServicePubSubTestFramework(publisher, subscriber)

  override def beforeAll(): Unit = {
    redis.start()
  }

  override def afterAll(): Unit = {
    redisClient.shutdown()
    redis.stop()
    wiring.shutdown(TestFinishedReason).await
  }

  test("Redis - should be able to publish and subscribe an event") {
    framework.pubSub()
  }

  test("Redis - should be able to make independent subscriptions") {
    framework.subscribeIndependently()
  }

  test("Redis - should be able to publish concurrently to the same channel") {
    framework.publishMultiple()
  }

  test("Redis - should be able to publish concurrently to the different channel") {
    framework.publishMultipleToDifferentChannels()
  }

  //DEOPSCSW-340: Provide most recently published event for subscribed prefix and name
  test("Redis - should be able to retrieve recently published event on subscription") {
    framework.retrieveRecentlyPublished()
  }

  //DEOPSCSW-340: Provide most recently published event for subscribed prefix and name
  test("Redis - should be able to retrieve InvalidEvent") {
    framework.retrieveInvalidEvent()
  }

  //DEOPSCSW-340: Provide most recently published event for subscribed prefix and name
  test("Redis - should be able to retrieve only valid events when one of the subscribed events keys has published events") {
    framework.retrieveMultipleSubscribedEvents()
  }

  //DEOPSCSW-338: Provide callback for Event alerts
  test("Redis - should be able to subscribe with callback") {
    framework.retrieveEventUsingCallback()
  }

  //DEOPSCSW-338: Provide callback for Event alerts
  test("Redis - should be able to subscribe with async callback") {
    framework.retrieveEventUsingAsyncCallback()
  }

  //DEOPSCSW-339: Provide actor ref to alert about Event arrival
  test("Redis - should be able to subscribe with an ActorRef") {
    framework.retrieveEventUsingActorRef()
  }

  //DEOPSCSW-344: Retrieve recently published event using prefix and eventname
  test("Redis - should be able to get an event without subscribing for it") {
    framework.get()
  }

  //DEOPSCSW-344: Retrieve recently published event using prefix and eventname
  test("Redis - should be able to get InvalidEvent") {
    framework.retrieveInvalidEventOnGet()
  }

  //DEOPSCSW-344: Retrieve recently published event using prefix and eventname
  test("Redis - should be able to get events for multiple event keys") {
    framework.retrieveEventsForMultipleEventKeysOnGet()
  }

  test("Redis - should be able to persist the event in DB against the same key as channel name while publishing") {
    val event1             = makeEvent(1)
    val eventKey: EventKey = event1.eventKey
    val redisCommands =
      redisClient.connect(EventServiceCodec, RedisURI.create(clusterSettings.hostname, redisPort)).sync()

    publisher.publish(event1).await

    Thread.sleep(1000)
    redisCommands.get(eventKey) shouldBe event1
  }
}

File: csw-event-impl/src/test/java/csw/services/event/redis/internal/PubSubTest.java
package csw.services.event.redis.internal;

import akka.actor.ActorSystem;
import akka.japi.Pair;
import akka.stream.javadsl.Keep;
import akka.stream.javadsl.Sink;
import csw.messages.commons.CoordinatedShutdownReasons;
import csw.messages.events.Event;
import csw.messages.events.Event$;
import csw.messages.events.EventKey;
import csw.services.event.JRedisFactory;
import csw.services.event.helpers.RegistrationFactory;
import csw.services.event.helpers.Utils;
import csw.services.event.internal.commons.EventServiceConnection;
import csw.services.event.internal.commons.Wiring;
import csw.services.event.javadsl.IEventPublisher;
import csw.services.event.javadsl.IEventSubscriber;
import csw.services.event.javadsl.IEventSubscription;
import csw.services.location.commons.ClusterAwareSettings;
import csw.services.location.commons.ClusterSettings;
import csw.services.location.models.TcpRegistration;
import csw.services.location.scaladsl.LocationService;
import csw.services.location.scaladsl.LocationServiceFactory;
import io.lettuce.core.RedisClient;
import org.junit.*;
import redis.embedded.RedisServer;
import scala.concurrent.Await;
import scala.concurrent.duration.FiniteDuration;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class PubSubTest {
    private static int seedPort = 3562;
    private static int redisPort = 6379;

    private static ClusterSettings clusterSettings = ClusterAwareSettings.joinLocal(seedPort);

    private static RedisServer redis = RedisServer.builder().setting("bind " + clusterSettings.hostname()).port(redisPort).build();
    private static RedisClient redisClient;
    private static Wiring wiring;
    private IEventPublisher publisher;
    private IEventSubscriber subscriber;

    public PubSubTest() throws Exception {
        TcpRegistration tcpRegistration = RegistrationFactory.tcp(EventServiceConnection.value(), redisPort);
        LocationService locationService = LocationServiceFactory.withSettings(ClusterAwareSettings.onPort(seedPort));
        Await.result(locationService.register(tcpRegistration), new FiniteDuration(10, TimeUnit.SECONDS));
        redisClient = RedisClient.create();
        ActorSystem actorSystem = clusterSettings.system();
        wiring = new Wiring(actorSystem);
        JRedisFactory redisFactory = new JRedisFactory(redisClient, locationService, wiring);
        publisher = redisFactory.publisher().get();
        subscriber = redisFactory.subscriber().get();
    }

    @BeforeClass
    public static void beforeClass() {
        redis.start();
    }

    @AfterClass
    public static void afterClass() throws Exception {
        redisClient.shutdown();
        redis.stop();
        Await.result(wiring.shutdown(CoordinatedShutdownReasons.TestFinishedReason$.MODULE$), new FiniteDuration(10, TimeUnit.SECONDS));
    }

    @Test
    public void shouldBeAbleToPublishAndSubscribeAnEvent() throws InterruptedException, TimeoutException, ExecutionException {
        Event event1 = Utils.makeDistinctEvent(1);
        EventKey eventKey = event1.eventKey();

        java.util.Set<EventKey> set = new HashSet<>();
        set.add(eventKey);

        Pair<IEventSubscription, CompletionStage<List<Event>>> subscriptionFuturePair = subscriber.subscribe(set).toMat(Sink.seq(), Keep.both()).run(wiring.resumingMat());
        Thread.sleep(2000);

        publisher.publish(event1);
        Thread.sleep(1000);

        subscriptionFuturePair.first().unsubscribe().get(10, TimeUnit.SECONDS);

        List<Event> expectedEvents = new ArrayList<>();
        expectedEvents.add(Event$.MODULE$.invalidEvent());
        expectedEvents.add(event1);
        Assert.assertArrayEquals(subscriptionFuturePair.second().toCompletableFuture().get().toArray(), expectedEvents.toArray());
    }
}
